integration/nwo/channel_participation.go

```diff
@@ -0,0 +1,171 @@
+	blockBytes, err := proto.Marshal(block)
+	Expect(err).NotTo(HaveOccurred())
+	url := fmt.Sprintf("https://127.0.0.1:%d/participation/v1/channels/%s", n.OrdererPort(o, OperationsPort), channel)
+	req := generateJoinRequest(url, channel, blockBytes)
+	authClient, _ := OrdererOperationalClients(n, o)
+
+	By("joining channel " + expectedChannelInfo.Name)
+	body := doBody(authClient, req)
+	c := &ChannelInfo{}
+	err = json.Unmarshal(body, c)
+	Expect(err).NotTo(HaveOccurred())
+	Expect(*c).To(Equal(expectedChannelInfo))
+}
+
+func generateJoinRequest(url, channel string, blockBytes []byte) *http.Request {
+	joinBody := new(bytes.Buffer)
+	writer := multipart.NewWriter(joinBody)
+	part, err := writer.CreateFormFile("config-block", fmt.Sprintf("%s.block", channel))
+	Expect(err).NotTo(HaveOccurred())
+	part.Write(blockBytes)
+	err = writer.Close()
+	Expect(err).NotTo(HaveOccurred())
+
+	req, err := http.NewRequest(http.MethodPost, url, joinBody)
+	Expect(err).NotTo(HaveOccurred())
+	req.Header.Set("Content-Type", writer.FormDataContentType())
+
+	return req
+}
+
+func doBody(client *http.Client, req *http.Request) []byte {
+	resp, err := client.Do(req)
+	Expect(err).NotTo(HaveOccurred())
+	Expect(resp.StatusCode).To(Equal(http.StatusCreated))
+	bodyBytes, err := ioutil.ReadAll(resp.Body)
+	Expect(err).NotTo(HaveOccurred())
+	resp.Body.Close()
+
+	return bodyBytes
+}
+
+type channelList struct {
+	SystemChannel *channelInfoShort  `json:"systemChannel"`
+	Channels      []channelInfoShort `json:"channels"`
+}
+
+type channelInfoShort struct {
+	Name string `json:"name"`
+	URL  string `json:"url"`
+}
+
+func ChannelParticipationList(n *Network, o *Orderer, expectedChannels []string, systemChannel ...string) {
+	authClient, unauthClient := OrdererOperationalClients(n, o)
+	listChannelsURL := fmt.Sprintf("https://127.0.0.1:%d/participation/v1/channels", n.OrdererPort(o, OperationsPort))
+
+	By("listing the channels")
+	body := GetBody(authClient, listChannelsURL)()
+	list := &channelList{}
+	err := json.Unmarshal([]byte(body), list)
+	Expect(err).NotTo(HaveOccurred())
+
+	Expect(*list).To(MatchFields(IgnoreExtras, Fields{
+		"Channels":      channelsMatcher(expectedChannels),
+		"SystemChannel": systemChannelMatcher(systemChannel...),
+	}))
+
+	By("listing the channels without a client cert")
+	resp, err := unauthClient.Get(listChannelsURL)
+	Expect(err).NotTo(HaveOccurred())
+	Expect(resp.StatusCode).To(Equal(http.StatusUnauthorized))
+
+	// for the current test scenarios, we only have these two cluster relations.
+	// other possible values are "follower" and "config-tracker"
+	// TODO generalize and relocate this after join operations are usable/testable
+	clusterRelation := func(consensusType string) string {
+		if consensusType == "etcdraft" {
+			return "member"
+		}
+		return "none"
+	}
+
+	for _, channel := range list.Channels {
```

Comment:
I think the time is ripe to take this out of the "List all" and maybe into a separate function that gets the result of the list, and an array of expected ChannelInfo, (i.e. expected []ChannelInfo), etc? or maybe a map channelName to ChannelInfo?

This way it would be possible to check `list all` multiple times as TXs come in, and the height, status, and clusterRelation change...

Reply:
Removed this logic.